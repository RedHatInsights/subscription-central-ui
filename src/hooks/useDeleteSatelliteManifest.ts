import Cookies from 'js-cookie';
import { useMutation, useQueryClient } from 'react-query';
import { ManifestEntry } from './useSatelliteManifests';

const deleteSatelliteManifest = (uuid: string) => {
  const jwtToken = Cookies.get('cs_jwt');
  return fetch(`/api/rhsm/v2/manifests/${uuid}?force=true`, {
    method: 'DELETE',
    headers: { Authorization: `Bearer ${jwtToken}` }
  }).then((response) => {
    if (response.status != 204) {
      throw new Error(
        `Status Code ${response.status}.  Error deleting manifest: ${response.statusText}.`
      );
    }
  });
};

const useDeleteSatelliteManifest = () => {
  const queryClient = useQueryClient();
  return useMutation((uuid: string) => deleteSatelliteManifest(uuid), {
    onSuccess: (data, uuid) => {
      queryClient.setQueryData('manifests', (oldData: ManifestEntry[]) =>
        oldData.filter((entry) => entry.uuid != uuid)
      );
    }
  });
};

export { deleteSatelliteManifest, useDeleteSatelliteManifest as default };
