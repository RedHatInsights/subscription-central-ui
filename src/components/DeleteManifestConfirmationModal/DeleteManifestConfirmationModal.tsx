import React, { FunctionComponent } from 'react';
import {
  Button,
  Checkbox,
  EmptyState,
  EmptyStateIcon,
  Modal,
  ModalVariant,
  Spinner,
  Text,
  TextContent,
  TextVariants,
  EmptyStateHeader
} from '@patternfly/react-core';
import useDeleteSatelliteManifest from '../../hooks/useDeleteSatelliteManifest';
import useNotifications from '../../hooks/useNotifications';

interface DeleteManifestConfirmationModalProps {
  handleModalToggle: () => void;
  onSuccess: () => void;
  isOpen: boolean;
  name: string;
  uuid: string;
}

const DeleteManifestConfirmationModal: FunctionComponent<DeleteManifestConfirmationModalProps> = ({
  handleModalToggle,
  onSuccess,
  isOpen,
  name,
  uuid
}) => {
  const [checked, setChecked] = React.useState(false);
  const { addSuccessNotification, addErrorNotification } = useNotifications();
  const {
    isError: manifestFailedToDelete,
    isLoading: isDeletingManifest,
    isSuccess: manifestDeleted,
    mutate: deleteManifest,
    reset: resetRequestStatus
  } = useDeleteSatelliteManifest();
  const requestCompleted = manifestDeleted || manifestFailedToDelete;

  const handleCheckbox = (event: React.ChangeEvent<HTMLInputElement>) => {
    setChecked(event.target.checked);
  };

  const closeModal = () => {
    handleModalToggle();
    setChecked(false);
  };

  const resetModal = () => {
    closeModal();
    resetRequestStatus();
  };

  const actions = () => {
    if (isDeletingManifest) {
      return [];
    } else {
      return [
        <Button
          key="confirm"
          variant="danger"
          isDisabled={!checked}
          onClick={() => deleteManifest(uuid)}
        >
          Delete
        </Button>,
        <Button key="cancel" variant="link" onClick={closeModal}>
          Cancel
        </Button>
      ];
    }
  };

  const Content = () => {
    if (isDeletingManifest) {
      return (
        <EmptyState>
          <EmptyStateHeader icon={<EmptyStateIcon icon={Spinner} />} />
        </EmptyState>
      );
    } else {
      return (
        <TextContent>
          <Text component={TextVariants.p}>
            <b>{name}</b> will be permanently deleted, and the connection to Red Hat Insights will
            be broken. To enable again, you must create, export, download, and import a new manifest
            into your connected Satellite Server.
          </Text>
          <Checkbox
            label="I acknowledge that this action cannot be undone"
            isChecked={checked}
            id="confirmation_checkbox"
            onChange={(event: React.ChangeEvent<HTMLInputElement>) => handleCheckbox(event)}
          />
        </TextContent>
      );
    }
  };

  if (manifestDeleted) {
    onSuccess();
    addSuccessNotification(`Manifest ${name} deleted`);
  } else if (manifestFailedToDelete) {
    addErrorNotification('Something went wrong. Please try again');
  }
  if (requestCompleted) resetModal();

  return (
    <Modal
      isOpen={isOpen}
      onClose={closeModal}
      title={isDeletingManifest ? `Deleting manifest...` : `Delete manifest?`}
      variant={ModalVariant.small}
      titleIconVariant="warning"
      actions={actions()}
    >
      <Content />
    </Modal>
  );
};

export default DeleteManifestConfirmationModal;
