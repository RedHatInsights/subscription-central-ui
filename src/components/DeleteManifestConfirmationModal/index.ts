import DeleteManifestConfirmationModal from './DeleteManifestConfirmationModal';
export default DeleteManifestConfirmationModal;
