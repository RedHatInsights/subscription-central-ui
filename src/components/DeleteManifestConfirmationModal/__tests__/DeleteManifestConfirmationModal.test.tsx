import React from 'react';
import { render } from '@testing-library/react';
import DeleteManifestConfirmationModal from '../DeleteManifestConfirmationModal';
import useDeleteSatelliteManifest from '../../../hooks/useDeleteSatelliteManifest';

jest.mock('../../../hooks/useDeleteSatelliteManifest');

describe('DeleteConfirmationModal', () => {
  const props = {
    isOpen: true,
    uuid: '00000000-0000-0000-0000-000000000000',
    name: 'Manifest Name',
    handleModalToggle: (): any => undefined,
    onSuccess: (): any => undefined
  };
  let isLoading = false;

  beforeEach(() => {
    (useDeleteSatelliteManifest as jest.Mock).mockImplementation(() => ({
      isLoading: isLoading,
      isSuccess: false,
      isError: false,
      reset: (): any => undefined
    }));
  });

  it('renders correctly', () => {
    render(
      <div>
        <DeleteManifestConfirmationModal {...props} />
      </div>
    );
    expect(document.body).toMatchSnapshot();
  });

  describe('when performing a delete', () => {
    beforeAll(() => {
      isLoading = true;
    });

    it('renders a spinner', () => {
      render(<DeleteManifestConfirmationModal {...props} />);
      expect(document.body).toMatchSnapshot();
    });
  });
});
