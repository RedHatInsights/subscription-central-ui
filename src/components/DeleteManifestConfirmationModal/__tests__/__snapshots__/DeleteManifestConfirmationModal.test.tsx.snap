// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteConfirmationModal renders correctly 1`] = `
<body
  class="pf-c-backdrop__open"
>
  <div
    aria-hidden="true"
  >
    <div />
  </div>
  <div>
    <div
      class="pf-c-backdrop"
    >
      <div
        class="pf-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-2"
          aria-labelledby="pf-modal-part-1"
          aria-modal="true"
          class="pf-c-modal-box pf-m-warning pf-m-sm"
          data-ouia-component-id="OUIA-Generated-Modal-small-1"
          data-ouia-component-type="PF4/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-0"
          role="dialog"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
          <header
            class="pf-c-modal-box__header"
          >
            <h1
              class="pf-c-modal-box__title pf-m-icon"
              id="pf-modal-part-1"
            >
              <span
                class="pf-c-modal-box__title-icon"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 576 512"
                  width="1em"
                >
                  <path
                    d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                  />
                </svg>
              </span>
              <span
                class="pf-u-screen-reader"
              >
                Warning alert:
              </span>
              <span
                class="pf-c-modal-box__title-text"
              >
                Delete Manifest Name
              </span>
            </h1>
          </header>
          <div
            class="pf-c-modal-box__body"
            id="pf-modal-part-2"
          >
            <div
              class="pf-c-content"
            >
              <p
                class=""
                data-pf-content="true"
              >
                Deleting a manifest is STRONGLY discouraged. Deleting a manifest will:
              </p>
              <ul
                class="pf-c-list"
              >
                <li>
                  Delete all subscriptions that are attached to running hosts.
                </li>
                <li>
                  Delete all subscriptions attached to activation keys.
                </li>
                <li>
                  Disable Red Hat Insights.
                </li>
                <li>
                  Require you to upload the manifest and re-attach subscriptions to hosts and activation keys.
                </li>
              </ul>
              <p
                class=""
                data-pf-content="true"
              >
                <strong>
                  Caution: this operation is permanent and cannot be undone.
                </strong>
              </p>
              <div
                class="pf-c-check"
              >
                <input
                  aria-invalid="false"
                  class="pf-c-check__input"
                  id="confirmation_checkbox"
                  type="checkbox"
                />
                <label
                  class="pf-c-check__label"
                  for="confirmation_checkbox"
                >
                  Are you sure you want to delete this manifest?
                </label>
              </div>
            </div>
          </div>
          <footer
            class="pf-c-modal-box__footer"
          >
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-link"
              data-ouia-component-id="OUIA-Generated-Button-link-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              NO, CANCEL
            </button>
            <button
              aria-disabled="true"
              class="pf-c-button pf-m-danger pf-m-disabled"
              data-ouia-component-id="OUIA-Generated-Button-danger-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              disabled=""
              type="button"
            >
              YES, DELETE
            </button>
          </footer>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`DeleteConfirmationModal when performing a delete renders a spinner 1`] = `
<body
  class="pf-c-backdrop__open"
>
  <div
    aria-hidden="true"
  />
  <div>
    <div
      class="pf-c-backdrop"
    >
      <div
        class="pf-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-part-3"
          aria-labelledby="pf-modal-part-2"
          aria-modal="true"
          class="pf-c-modal-box pf-m-warning pf-m-sm"
          data-ouia-component-id="OUIA-Generated-Modal-small-2"
          data-ouia-component-type="PF4/ModalContent"
          data-ouia-safe="true"
          id="pf-modal-part-1"
          role="dialog"
        >
          <button
            aria-disabled="false"
            aria-label="Close"
            class="pf-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-2"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              />
            </svg>
          </button>
          <header
            class="pf-c-modal-box__header"
          >
            <h1
              class="pf-c-modal-box__title pf-m-icon"
              id="pf-modal-part-2"
            >
              <span
                class="pf-c-modal-box__title-icon"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 576 512"
                  width="1em"
                >
                  <path
                    d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                  />
                </svg>
              </span>
              <span
                class="pf-u-screen-reader"
              >
                Warning alert:
              </span>
              <span
                class="pf-c-modal-box__title-text"
              >
                Deleting Manifest Name...
              </span>
            </h1>
          </header>
          <div
            class="pf-c-modal-box__body"
            id="pf-modal-part-3"
          >
            <div
              class="pf-c-empty-state"
            >
              <div
                class="pf-c-empty-state__content"
              >
                <div
                  class="pf-c-empty-state__icon"
                >
                  <span
                    aria-valuetext="Loading..."
                    class="pf-c-spinner pf-m-xl"
                    role="progressbar"
                  >
                    <span
                      class="pf-c-spinner__clipper"
                    />
                    <span
                      class="pf-c-spinner__lead-ball"
                    />
                    <span
                      class="pf-c-spinner__tail-ball"
                    />
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
